### PR TITLE
Added GMRES solver

### PR DESCRIPTION
**Pull request by k7miller**
- Added GMRES iterative solver: GMRESSolver.hpp, GMRESSolver.cpp.
- Added testing code for GMRES: GMRESSolverTests.cpp.
- Added helper functions for basic linear algebra operations: LinAlgHelpers.hpp, LinAlgHelpers.cpp.

### CODE REVIEW

File: include/GMRESSolver.hpp
Comment (dsteinmo) [inline include/GMRESSolver.hpp]: I think I'd like to see namespaces start a new level of indentation.

File: include/GMRESSolver.hpp
Comment (dsteinmo) [inline include/GMRESSolver.hpp]: Nice documentation, can we change the commenting syntax to use the /**  */ style used elsewhere in the project? Otherwise doxygen won't know to pull these into the docs

File: src/test/GMRESSolverTests.cpp
Comment (dsteinmo) [inline src/test/GMRESSolverTests.cpp]: Super legit testing

File: src/test/GMRESSolverTests.cpp
Comment (dsteinmo) [inline src/test/GMRESSolverTests.cpp]: Why the random numbers? Is this for initial guess?

File: src/GMRESSolver.cpp
Comment (dsteinmo) [inline src/GMRESSolver.cpp]: Very cool use of operator overloads

**Review by dsteinmo — APPROVED** [review]
Looks amazing!  I approve, but let's hold off on merging until we sort out the segmentation fault in Travis.

File: src/test/GMRESSolverTests.cpp
Comment (dsteinmo) [inline src/test/GMRESSolverTests.cpp]: I think you need to capitalize the U in SetUp, but maybe check the other test files. This could be why it breaks in the Linux build (accessing something that's not allocated yet)

File: src/GMRESSolver.cpp
Comment (k7miller) [reply src/GMRESSolver.cpp]: Thanks bud!

File: src/test/GMRESSolverTests.cpp
Comment (k7miller) [reply src/test/GMRESSolverTests.cpp]: Yah, just wanted to test it with a nonzero initial guess.

File: include/GMRESSolver.hpp
Comment (k7miller) [reply include/GMRESSolver.hpp]: Sure thing bud.

File: include/GMRESSolver.hpp
Comment (k7miller) [reply include/GMRESSolver.hpp]: Roger that Frank, I will have to review the comments in the code to see how this is done. I am not familiar with Doxygen, so I will have to check that out too.

**Comment by dsteinmo** [comment]
Hey buddy, I hacked the build to stash the 'bin/test' binary somewhere and ran valgrind against it. It looks like blitz is having trouble figuring out the bounds of the input array in MatVec's operator that lives in the test class.

`==25705== Conditional jump or move depends on uninitialised value(s)
==25705==    at 0x425990: blitz::bounds::compute_ordering(int, int, int) (bounds.h:40)
==25705==    by 0x4D7B87: blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> >::ordering(int) const (expr.h:945)
==25705==    by 0x4D3F9B: blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> > >::ordering(int) const (expr.h:203)
==25705==    by 0x4D1C7E: blitz::_bz_ArrayExprReduce<blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> > >, 1, blitz::ReduceSum<double, double> >::computeOrdering() (reduce.h:290)
==25705==    by 0x4CF88B: blitz::_bz_ArrayExprReduce<blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> > >, 1, blitz::ReduceSum<double, double> >::_bz_ArrayExprReduce(blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> > >) (reduce.h:116)
==25705==    by 0x4CA8D6: blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprReduce<blitz::asExpr<blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> > > >::T_expr, 1, blitz::ReduceSum<blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> > >::T_numtype, blitz::NumericTypeTraits<blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> > >::T_numtype>::T_sumtype> > > blitz::sum<blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> > >, 1>(blitz::ETBase<blitz::_bz_ArrayExpr<blitz::_bz_ArrayExprBinaryOp<blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 2> >, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::_bz_ArrayExpr<blitz::ArrayIndexMapping<blitz::_bz_ArrayExpr<blitz::FastArrayIterator<double, 1> >, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0> >, blitz::Multiply<double, double> > > > const&, blitz::IndexPlaceholder<1> const&) (in /home/dsteinmo/blitzdg/bin/test)
==25705==    by 0x4C72F1: GMRESSolverTests::Matvec::operator()(blitz::Array<double, 1> const&, blitz::Array<double, 1>&) const (GMRESSolverTests.cpp:21)
==25705==    by 0x4CB8B1: blitzdg::GMRESOut blitzdg::GMRESSolver::solve<GMRESSolverTests::Matvec, GMRESSolverTests::Precon>(GMRESSolverTests::Matvec, GMRESSolverTests::Precon, blitz::Array<double, 1> const&, blitz::Array<double, 1>&, blitzdg::GMRESParams const&) const (GMRESSolver.hpp:143)
==25705==    by 0x4C859F: GMRESSolverTests::GMRESSolver_Object::Solve_Ax_equals_b() (GMRESSolverTests.cpp:72)
==25705==    by 0x4E6F5A: igloo::ContextRegistry<GMRESSolverTests::GMRESSolver_Object>::CallSpec(GMRESSolverTests::GMRESSolver_Object&, std::string const&, void (GMRESSolverTests::GMRESSolver_Object::*)(), igloo::TestResults&) (contextregistry.h:125)
==25705==    by 0x4E65B3: void igloo::ContextRegistry<GMRESSolverTests::GMRESSolver_Object>::CallSpecs<GMRESSolverTests::GMRESSolver_Object>(std::map<std::string, igloo::ContextRegistry<GMRESSolverTests::GMRESSolver_Object>::SpecInfo, std::less<std::string>, std::allocator<std::pair<std::string const, igloo::ContextRegistry<GMRESSolverTests::GMRESSolver_Object>::SpecInfo> > > const&, std::string const&, igloo::TestResults&, igloo::TestListener&) (contextregistry.h:101)
==25705==    by 0x4E5E6E: void igloo::ContextRegistry<GMRESSolverTests::GMRESSolver_Object>::Run<GMRESSolverTests::GMRESSolver_Object>(std::string const&, igloo::TestResults&, igloo::TestListener&) (contextregistry.h:73)`

The run time that's on travis might be being extra picky about something we're not catching locally, so it might be worth trying to refactor things so that the A matrix's memory lives on the operator (and not just a reference). I don't know exactly why this way isn't working though, just trying to see what's different here...

**Comment by dsteinmo** [comment]
Looks like Travis and AppVeyor agree with my changes. I'll let you look it over, and if you're cool with my refactor, let's ship it to master :shipit: .

File: src/test/GMRESSolverTests.cpp
Comment (k7miller) [reply src/test/GMRESSolverTests.cpp]: I believe that blitz uses reference semantics for copy construction. So, the member variable Matvec::mat_ should be created as an alias of the input matrix. Because of this behavior, I don't think we need to worry about any overhead incurred by passing the Matvec functor by value. Clearly, this isn't an issue here, since A is only 5 x 5, but it would be good to know in general. There is also the reference() function in blitz, which may do the same thing.

**Pull request merged**
